internal/curator/durable/state/fb/extra.go

```diff
@@ -0,0 +1,215 @@
+		(uint32(repl)&0xff)<<metaRepl
+}
+
+// Expand/truncate unix nanos/seconds so we can work with nanos in the rest of
+// the codebase:
+
+func (b *BlobF) Deleted() int64 { return int64(b.Deletedsec()) * 1e9 }
+func (b *BlobF) Mtime() int64   { return int64(b.Mtimesec()) * 1e9 }
+func (b *BlobF) Atime() int64   { return int64(b.Atimesec()) * 1e9 }
+func (b *BlobF) Expires() int64 { return int64(b.Expiressec()) * 1e9 }
+
+func (b *BlobF) MutateDeleted(n int64) bool { return b.MutateDeletedsec(uint32(n / 1e9)) }
+func (b *BlobF) MutateMtime(n int64) bool   { return b.MutateMtimesec(uint32(n / 1e9)) }
+func (b *BlobF) MutateAtime(n int64) bool   { return b.MutateAtimesec(uint32(n / 1e9)) }
+func (b *BlobF) MutateExpires(n int64) bool { return b.MutateExpiressec(uint32(n / 1e9)) }
+
+func BlobFAddDeleted(b *flatbuffers.Builder, n int64) { BlobFAddDeletedsec(b, uint32(n/1e9)) }
+func BlobFAddMtime(b *flatbuffers.Builder, n int64)   { BlobFAddMtimesec(b, uint32(n/1e9)) }
+func BlobFAddAtime(b *flatbuffers.Builder, n int64)   { BlobFAddAtimesec(b, uint32(n/1e9)) }
+func BlobFAddExpires(b *flatbuffers.Builder, n int64) { BlobFAddExpiressec(b, uint32(n/1e9)) }
+
+// Helpers for TractIDF:
+
+func (t *TractIDF) TractID() core.TractID {
+	bid := uint64(t.B5())<<48 | uint64(t.B4())<<32 | uint64(t.B3())<<16 | uint64(t.B2())
+	return core.TractID{Blob: core.BlobID(bid), Index: core.TractKey(t.B1())}
+}
+
+func (t *TractIDF) RSChunkID() core.RSChunkID {
+	pid := uint32(t.B5())<<16 | uint32(t.B4())
+	id := uint64(t.B3())<<32 | uint64(t.B2())<<16 | uint64(t.B1())
+	return core.RSChunkID{Partition: core.PartitionID(pid), ID: id}
+}
+
+func PutTractID(bu *flatbuffers.Builder, t core.TractID) flatbuffers.UOffsetT {
+	return CreateTractIDF(bu,
+		uint16(t.Index),
+		uint16(t.Blob>>0),
+		uint16(t.Blob>>16),
+		uint16(t.Blob>>32),
+		uint16(t.Blob>>48))
+}
+
+func PutRSChunkID(bu *flatbuffers.Builder, c core.RSChunkID) flatbuffers.UOffsetT {
+	return CreateTractIDF(bu,
+		uint16(c.ID>>0),
+		uint16(c.ID>>16),
+		uint16(c.ID>>32),
+		uint16(c.Partition>>0),
+		uint16(c.Partition>>16))
+}
+
+// Helpers for RSC_TractF:
+
+func (t *RSC_TractF) TractID() core.TractID {
+	var tid TractIDF
+	return t.Id(&tid).TractID()
+}
+
+func CreateRSC_TractFWithID(bu *flatbuffers.Builder, t core.TractID, length uint32, offset uint32) flatbuffers.UOffsetT {
+	return CreateRSC_TractF(bu,
+		uint16(t.Index),
+		uint16(t.Blob>>0),
+		uint16(t.Blob>>16),
+		uint16(t.Blob>>32),
+		uint16(t.Blob>>48),
+		length, offset)
+}
+
+// Helpers for TractF:
+
+// HostsLength makes Hosts in TractF look like a normal field.
+func (t *TractF) HostsLength() (count int) {
+	hosts012 := t.Hosts012()
+	if hosts012&(0xfffff<<0) == 0 {
+		return 0
+	}
+	if hosts012&(0xfffff<<20) == 0 {
+		return 1
+	}
+	if hosts012&(0xfffff<<40) == 0 {
+		return 2
+	}
+	return 3 + t.Hosts3pLength()
+}
+
+// Hosts makes Hosts in TractF look like a normal field.
+func (t *TractF) Hosts(j int) uint32 {
```

Comment:
nit: Extracting 3p hosts means j should be >= 3, -1 -9999 will default to 3p case right? Check this for defense in depth?

Reply:
Flatbuffers basically does no range checking in its accessors, it just lets Go panic if you end up out of bounds of the buffer. I wrote these to match that behavior, since it would feel a little weird if this did range checking but the official accessors didn't. It just assumes that you're going to call HostsLength and only use values in [0, length).

A related question is about verifying the buffer so that you can be sure that it won't panic when used correctly. Other flatbuffers implementation languages include a verifier, but the Go support doesn't. In general, since we're the only thing writing to the db, we can assume that everything in the db is properly formed. But like the consistency check, it'd be nice to verify things. If someone implements verification support in Go, we should probably call it sometimes, either some fraction of the time when data is read, or in a background loop.

Reply:
Good to know. Scratch my comment here then.